UnitySetup/UnitySetup.psm1

```diff
@@ -2132,3 +2130,537 @@ function Get-UnityLicense {
+                Auth = $convertedScopedPAT
+            }
+        }
+        $Results += [PSCustomObject]@{
+            Scoped_URL = $scopedRegistryURL
+            Auth_State = $AuthState
+        }
+    }
+
+    return $UPMConfigs
+}
+
+function Export-UPMConfig {
+    [CmdletBinding()]
+    param(
+        [PSCustomObject[]]$UPMConfigs,
+        [string[]]$tomlFilePaths
+    )
+
+    foreach ($UPMConfig in $UPMConfigs) {
+        if (![string]::IsNullOrEmpty($UPMConfig.Scoped_URL) -and ![string]::IsNullOrEmpty($UPMConfig.Auth)) {
+            $scopedRegistryURL = $UPMConfig.Scoped_URL
+            $convertedScopedPAT = $UPMConfig.Auth
+
+            $tomlConfigContent = @(
+                "`r`n[npmAuth.""$scopedRegistryURL""]"
+                "_auth = ""$convertedScopedPAT"""
+                "alwaysAuth = true"
+            ) -join "`r`n"
+
+            foreach ($filePath in $tomlFilePaths) {
+                Add-Content -Path $filePath -Value $tomlConfigContent
+            }
+        }
+    }
+}
+
+<#
+.Synopsis
+   Ensures that the user has the appropriate auth tokens to fetch Unity packages in their .toml file.
+
+   For more information on Unity Package Manager config, please visit https://docs.unity3d.com/Manual/upm-config.html
+.DESCRIPTION
+   Looks at the Unity Project Manifest and finds the scoped registries used for fetching NPM packages.  
+
+   For each of the scoped registries found within the project manifest(s) or SOT.json file, the cmdlet will verify that
+   there is a valid auth token for each scoped registry URL.  If none were found, it will try to fetch a new auth token
+   and save it to the .toml file.
+
+   Additional arguments are available to automatically clean expired tokens, allow the user to manually propulate the token,
+   scan a deeper folder tree for manifests, or simply validate your existing PATs.
+.PARAMETER ProjectManifestPath
+   A path to a project manifest, or a path to a root directory under which Unity project manifests can be found.
+.PARAMETER AutoClean
+   Automatically remove PATs that can't be validated
+.PARAMETER NoValidation
+   Skip validation of PATs
+.PARAMETER ManualPAT
+   Do not use Azure APIs to automatically create the PAT, user will manually enter it
+.PARAMETER SearchDepth
+   How deep to search for manifest files
+.PARAMETER VerifyOnly
+   Runs in validation only mode, returns 0 if all registries are valid, otherwise returns 1
+.PARAMETER PATLifetime
+   How many days the created PAT is valid
+.EXAMPLE
+   Update-UPMConfig -ProjectManifestPath '/User/myusername/MyUnityProjectRoot'
+.EXAMPLE
+   Update-UPMConfig -ProjectManifestPath '/User/myusername/MyUnityProjectRoot/manifest.json'
+.EXAMPLE
+   Update-UPMConfig -AutoClean True
+.EXAMPLE
+   Update-UPMConfig -NoValidation True -ManualPAT True
+.EXAMPLE
+   Update-UPMConfig -ProjectManifestPath '/User/myusername/MyUnityProjectRoot' -SearchDepth 7 -VerifyOnly True
+#>
+function Update-UPMConfig {
```

Comment:
This method should take a upm config and a manifest (both loaded from Import cmdlets) and update the upm config such that it supports all the scoped registries in the manifest.

Reply:
Gotcha, I think I get the jist of the preferred structure you're suggesting.  I'll refactor the Import functions to pull the manifest / .toml (UMP Config) separately to pass this in.  Based on the other comment, I'll assume that the -ProjectManifest parameter can take either a path or take the output of a new `Import-ProjectManifest` cmdlet directly.

Reply:
I think maybe we have `ProjectManifest` which takes manifests loaded from e.g. `Import-UnityProjectManifest` (which I hope you'll be adding), but also maybe `ProjectManifestPath` which could be passed to `Import-UnityProjectManifest` as a helper to avoid users needing to do that first if they don't need to.

Reply:
So just to recap, internally the script supports 3 different modes.
- Directory target:  Scan and find every manifest to a given depth
- Manifest target:  Pass in a path to a specific manifest
- SOT target:  Pass in a path to a manifest equivalent (a super-manifest indexing all of the scopes)

Here's an example of run that does a scan:
![image](https://github.com/microsoft/unitysetup.powershell/assets/6227906/0bf8cd32-bd1a-40b3-acd1-038b6e95ada3)

Here's a run with a specific manifest path provided:
![image](https://github.com/microsoft/unitysetup.powershell/assets/6227906/124c6f4e-0c9a-4d6d-b95b-b3585f88cdb5)

Here's a run with a manifest-like target (SOT):
![image](https://github.com/microsoft/unitysetup.powershell/assets/6227906/048871c8-dbd4-4c2f-870f-de8277221276)

Realistically I think the most common usage will be to give it a target directory and just let the script find all of the manifests in the directory tree.  Do you think the current state of the script where we do an `Import-ProjectManifests` to import one or more manifests is all right for the approach?